### PR TITLE
F/zmq

### PR DESCRIPTION
**Pull request by nikhar-abbas**
## Description and Purpose
Enable zeroMQ interface - specifically set up for yaw control

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Examples/Testing, if applicable
Added example

## Checklist
- [x] Add example
- [x] Update docs with API changes

### CODE REVIEW

File: ROSCO/src/ReadSetParameters.f90 [hunk 1]
```diff
@@ -26,12 +26,15 @@ MODULE ReadSetParameters
...
+    SUBROUTINE ReadAvrSWAP(avrSWAP, LocalVar, zmqVar)
+        USE ROSCO_Types, ONLY : LocalVariables, ZMQ_Variables
 
         REAL(ReKi), INTENT(INOUT) :: avrSWAP(*)   ! The swap array, used to pass data to, and receive data from, the DLL controller.
         TYPE(LocalVariables), INTENT(INOUT) :: LocalVar
-        
+        TYPE(ZMQ_Variables),  INTENT(INOUT) :: zmqVar
+
+        real(C_DOUBLE), dimension(0:4) :: setpoints
+
```
Comment (dzalkind) [inline ROSCO/src/ReadSetParameters.f90:37]: Where are `zmqVar` and `setpoints` used?

File: ROSCO/src/ReadSetParameters.f90 [hunk 2]
```diff
@@ -388,12 +393,19 @@ SUBROUTINE ReadControlParameterFileSub(CntrPar, accINFILE, accINFILE_size,ErrVar
...
         CALL ParseInput(UnControllerParameters,CurLine,'DLL_FileName',accINFILE(1),CntrPar%DLL_FileName,ErrVar)
         CALL ParseInput(UnControllerParameters,CurLine,'DLL_InFile',accINFILE(1),CntrPar%DLL_InFile,ErrVar)
         CALL ParseInput(UnControllerParameters,CurLine,'DLL_ProcName',accINFILE(1),CntrPar%DLL_ProcName,ErrVar)
+        CALL ReadEmptyLine(UnControllerParameters,CurLine)   
+
+        !------------ ZeroMQ ------------
+        CALL ReadEmptyLine(UnControllerParameters,CurLine)   
+        CALL ParseInput(UnControllerParameters,CurLine,'ZMQ_CommAddress',accINFILE(1), zmqVar%ZMQ_CommAddress,ErrVar)
+		CALL ParseInput(UnControllerParameters,CurLine,'ZMQ_updateFreq',accINFILE(1), zmqVar%ZMQ_UpdateFreq,ErrVar)
+		zmqVar%ZMQ_UpdateCounter = 99999999 ! Initialize as very large number
```
Comment (dzalkind) [inline ROSCO/src/ReadSetParameters.f90:408]: Do we want these to be `zmqVar`s or `CntrPar` like everything else?

File: ROSCO_toolbox/utilities.py [hunk 3]
```diff
@@ -153,17 +156,13 @@ def write_DISCON(turbine, controller, param_file='DISCON.IN', txt_filename='Cp_C
     file.write('{}              ! WE_FOPoles        - First order system poles [1/s]\n'.format(''.join('{:<10.8f} '.format(rosco_vt['WE_FOPoles'][i]) for i in range(len(rosco_vt['WE_FOPoles'])))))
     file.write('\n')
     file.write('!------- YAW CONTROL ------------------------------------------------------\n')
-    file.write('{:<13.5f}       ! Y_ErrThresh		- Yaw error threshold. Turbine begins to yaw when it passes this. [rad^2 s]\n'.format(rosco_vt['Y_ErrThresh']))
+    file.write('{:<13.5f}       ! Y_uSwitch		- Integrator saturation (maximum signal amplitude contribution to pitch from yaw-by-IPC), [rad]\n'.format(rosco_vt['Y_uSwitch']))
+    file.write('{}! Y_ErrThresh    - Yaw error threshold. Turbine begins to yaw when it passes this. If Y_uSwitch is zero, only the first value is used. [deg].\n'.format(''.join('{:<4.6f}  '.format(rosco_vt['Y_ErrThresh'][i]) for i in range(len(rosco_vt['F_FlCornerFreq'])))))
```
Comment (dzalkind) [inline ROSCO_toolbox/utilities.py:160]: Are there multiple values here? If so, what is the second value?

File: ROSCO_toolbox/utilities.py [hunk 4]
```diff
@@ -207,8 +206,12 @@ def write_DISCON(turbine, controller, param_file='DISCON.IN', txt_filename='Cp_C
     file.write('"{}"            ! DLL_FileName        - Name/location of the dynamic library in the Bladed-DLL format\n'.format(rosco_vt['DLL_FileName']))
     file.write('"{}"            ! DLL_InFile          - Name of input file sent to the DLL (-)\n'.format(rosco_vt['DLL_InFile']))
     file.write('"{}"            ! DLL_ProcName        - Name of procedure in DLL to be called (-) \n'.format(rosco_vt['DLL_ProcName']))    
+    file.write('\n')
+    file.write('!------- ZeroMQ Interface ---------------------------------------------------------\n')
+    file.write('"{}"            ! ZMQ_CommAddress     - Communication address for ZMQ server \n'.format(rosco_vt['ZMQ_CommAddress']))
```
Comment (dzalkind) [inline ROSCO_toolbox/utilities.py:211]: What format is this supposed to be? Is this a port address?  An example would be pretty helpful.

File: ROSCO_toolbox/utilities.py [hunk 5]
```diff
@@ -207,8 +206,12 @@ def write_DISCON(turbine, controller, param_file='DISCON.IN', txt_filename='Cp_C
...
+    file.write('{:<11d}         ! ZMQ_updateFreq      - Call ZeroMQ every [x] timesteps, [integer]\n'.format(int(rosco_vt['ZMQ_updateFreq'])))
```
Comment (dzalkind) [inline ROSCO_toolbox/utilities.py:212]: If I'm being nitpickly, this is an update period, and it should be ZMQ_UpdateFreq to follow the same convention.

**Review by dzalkind — CHANGES_REQUESTED** [review]
Great work and contribution overall.  Would be nice to demo the new yaw controller at some point.  Can you clean up the few things I pointed out?

File: ROSCO/src/ReadSetParameters.f90 (hunk 2 above)
Comment (nikhar-abbas) [reply ROSCO/src/ReadSetParameters.f90]: Good catch - I moved these to `CntrPar`

File: ROSCO_toolbox/utilities.py (hunk 3 above)
Comment (nikhar-abbas) [reply ROSCO_toolbox/utilities.py:160]: There can be two values for `Y_uSwitch`. I fixed the `Y_ErrThresh` description to describe what happens. 

If `Y_uSwitch` > 0, the first value of `Y_ErrThresh` will be used for u < `Y_uSwitch`, and the second value will be used for u >= `Y_uSwitch`.

File: ROSCO_toolbox/utilities.py (hunk 4 above)
Comment (nikhar-abbas) [reply ROSCO_toolbox/utilities.py]: Added an example in the variable description

File: ROSCO_toolbox/utilities.py (hunk 5 above)
Comment (nikhar-abbas) [reply ROSCO_toolbox/utilities.py]: Nothin' wrong with being nitpickly.

I updated this to be `ZMQ_UpdatePeriod`, and it is now a time-period by which to update the ZMQ Coms (e.g. every 2 seconds).

**Review by dzalkind — APPROVED** [review]

**Pull request merged**